src/components/tl/cuda/reduce_scatterv/reduce_scatterv_linear.c

```diff
@@ -0,0 +1,450 @@
+ *
+ * Description:
+ *      scratch buffer is split into 2 parts to guarantee data consistency e.g.
+ *      when ranks in the ring are running at different steps of the algorithm.
+ *      2 parts is enough since max difference between ranks steps is 1
+ *
+ *  Definitions:
+ *      blockI  - full send buffer at Rank I
+ *      fragI_J - fragment of send buffer at Rank I and step J
+ *      NS      - number of steps
+ *      NF      - number of fragments
+ *      N       - team size
+ *
+ *  Setup:
+ *      max_frag_size = ucc_min(ucc_max(block0, block1, ..., block N -1 ),
+ *                              scratch_size / 2 / N)
+ *      NF            = ucc_max(block0, block1, ..., block N-1) / max_frag_size
+ *      NS            = 1 + NF
+ *
+ *  Algorithm
+ *      for rank R
+ *      step 0:    copy fragR_0 to remote scratch buffers for all ranks
+ *
+ *      step 1:    reduce frag1_0, frag2_0, ..., fragN_0 from local scratch buffer
+ *                 to local dst buffer
+ *                 copy fragR_1 from local src buffer to remote scratch buffers
+ *                 for all ranks
+ *      ...
+ *
+ *      step NS-2: reduce frag1_(NS-3), frag2_(NS-3), ..., fragN_(NS-3) from local
+ *                 scratch buffer to local dst buffer
+ *                 copy fragR_(NS-2) from local src buffer to remote scratch buffers
+ *                 for all ranks
+ *
+ *      step NS-1: reduce frag1_(NS-2), frag2_(NS-2), ..., fragN_(NS-2) from local
+ *                 scratch buffer to local dst buffer
+ */
+
+enum
+{
+    STAGE_SYNC,    /*< Wait for free SYNC segment */
+    STAGE_SETUP,   /*< Wait for memhandle setup to finish */
+    STAGE_COPIES,  /*< Linear algorithm is running */
+    STAGE_BARRIER, /*< Linear algorithm is done, waiting for
+                    *  other ranks to finish */
+};
+
+static inline int get_rank_step(ucc_tl_cuda_task_t *task, ucc_rank_t rank,
+                                int step_id)
+{
+    ucc_tl_cuda_sync_t *sync = TASK_SYNC(task, rank);
+
+    return sync->seq_num[step_id];
+}
+
+static inline void set_rank_step(ucc_tl_cuda_task_t *task, ucc_rank_t rank,
+                                 int step, int step_id)
+{
+    ucc_tl_cuda_sync_t *sync = TASK_SYNC(task, rank);
+
+    sync->seq_num[step_id] = step;
+}
+
+ucc_status_t
+ucc_tl_cuda_reduce_scatterv_linear_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+
+    tl_trace(UCC_TASK_LIB(task), "finalizing task %p", task);
+    ucc_tl_cuda_task_put(task);
+    return UCC_OK;
+}
+
+/*
+ * get maximum size of scratch memory evenly divisble by
+ * (2 * team_size * datatype_size)
+ */
+static inline size_t get_scratch_size(ucc_tl_cuda_team_t *team,
+                                      ucc_datatype_t      dt)
+{
+    size_t     dt_size = ucc_dt_size(dt);
+    ucc_rank_t tsize   = UCC_TL_TEAM_SIZE(team);
+    size_t     div     = 2 * dt_size * tsize;
```

Comment:
i'd suggest to rename local var "div" since there is a C func with the same name. Call it, e.g., divisor